File: java-spring/src/main/java/org/burgas/javaspring/service/CrudService.java
package org.burgas.javaspring.service;

import org.burgas.javaspring.dto.Request;
import org.burgas.javaspring.dto.Response;
import org.burgas.javaspring.entity.Entity;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.UUID;

@Component
public interface CrudService<R extends Request, E extends Entity, S extends Response, F extends Response> {

    E findEntity(UUID id);

    F findById(UUID id);

    List<S> findAll();

    F create(R request);

    F update(R request);

    void delete(UUID id);
}

File: kotlin-spring/src/main/kotlin/org/burgas/kotlinspring/service/CrudService.kt
package org.burgas.kotlinspring.service

import org.burgas.kotlinspring.entity.Model
import org.burgas.kotlinspring.entity.Request
import org.burgas.kotlinspring.entity.Response
import org.springframework.stereotype.Service
import java.util.UUID

@Service
interface CrudService<R : Request, M : Model, S : Response, F : Response> {

    fun findEntity(id: UUID): M

    fun findAll(): List<S>

    fun findById(id: UUID): F

    fun create(request: R)

    fun update(request: R)

    fun delete(id: UUID)
}